Add custom format support for hilog

xymeng16 opened the pull request:
This pull request introduces custom format support for HiLog 

- The interface and usage is fully compatible with `env_logger`'s, except color support due to the limitation of hilog itself.
- The implementation is mostly based on `env_logger`'s, with a custom `HilogWriter` that routes formatted logs to the hilog C interface.

## Usage Example
Here's an example of how to use the new custom formatting feature:

```rust
use hilog::{Builder, LogDomain};
use log::LevelFilter;
use std::io::Write;

let mut builder = Builder::new();
builder
    .set_domain(LogDomain::new(0x0000))
    .filter_level(LevelFilter::Trace)
    .filter_module("homoapp_native", LevelFilter::Trace)
    .format(|buf, record| {
        writeln!(
            buf,
            "{}:{} {} [{}] - {}",
            record.file().unwrap_or("unknown"),
            record.line().unwrap_or(0),
            chrono::Local::now().format("%Y-%m-%dT%H:%M:%S"),
            record.level(),
            record.args()
        )
    });
builder.init();
```

jschwe commented:
I think we definetly want an option for the user to add a custom formatter - but I don't think we should base our solution too strongly on `env_logger`, since some things don't apply to hilog. Instead I would recommend looking at android logger: https://github.com/rust-mobile/android_logger-rs/blob/master/src/lib.rs

Things to keep in mind: 
- hilog already logs the timestamp, so we don't need to format it into the message
- hilog already logs the log-level, so we don't need to format it into the message.
- hilog has a tag field, that the user might want to customise
- Both tag and log message have a maximum length (I believe 30 bytes (+ `\0`) for the tag and 4000 bytes for the message, so we might need to split a long log message into multiple hilog calls.
- I think we should avoid `ThreadLocal` and instead either allocate completely on the stack (like the C++ version of hilog does), or just do a heap allocation and use `format!()` for the message. I've previously experienced some bad performance of thread local on OH.

android_logger seems to basically already do almost everything we need, so I would recommend basing your solution on android_logger, which is also MIT and Apache2 licensed.

xymeng16 commented:
> but I don't think we should base our solution too strongly on env_logger, since some things don't apply to hilog.

I agree with your opinion that env_logger's implementation is somehow overpowered. And I will take a look at the `android_logger-rs`.

> - hilog already logs the timestamp, so we don't need to format it into the message
> - hilog already logs the log-level, so we don't need to format it into the message.
> - hilog has a tag field, that the user might want to customise

I have awared those points when porting the `env_logger`'s implementation. But the format is decided by the user, and In think we should not interfere too much on this. So on my own side, I don't think we need to add any limitations to the format it self.

> - Both tag and log message have a maximum length (I believe 30 bytes (+ \0) for the tag and 4000 bytes for the message, so we might need to split a long log message into multiple hilog calls.

This will be carefully handled in the next commit. HarmonyOS Nexy's document claims that the message is at most 4096 bytes and the exceeded part will be truncated. But I haven't find the limitation of the tag length yet. If you know any source defining them, please share me a link.

jschwe commented:
> But the format is decided by the user, and In think we should not interfere too much on this. So on my own side, I don't think we need to add any limitations to the format it self.

I agree, the custom format function provided by the user can of course do whatever they want. But I think it doesn't make sense for the **default** format function to have options to log the time and level.

>  But I haven't find the limitation of the tag length yet. If you know any source defining them, please share me a link.

You can view the libhilog implementation here:
https://gitee.com/openharmony/hiviewdfx_hilog/blob/master/frameworks/libhilog/include/hilog_base.h

xymeng16 committed:
feat: introduce custom format and message spliting

This commit implements a `HiLogWriter` that redirects everything it
receives to the `hilog` ffi-binding. It takes arbitrarily long message
and automatically splits it into several short message within the length
limitation of `hilog`, and then, sends them to the `hilog` binding.

xymeng16 commented:
This update removes the previous overpowered implementation, as discussed, and aligns with the simpler logic of `android_logger-rs`.

I’ve added two public interfaces to the Builder implementation:

```rust
// If tag is not set, the module path will be used.
pub fn set_tag(&mut self, tag: &str) -> &mut Self

// If custom format is not provided, raw message will be sent to `hilog`.
pub fn format<F>(&mut self, format: F) -> &mut Self
    where
        F: Fn(&mut dyn fmt::Write, &Record) -> fmt::Result + Sync + Send + 'static
```

xymeng16 committed:
chore: remove unused imports

xymeng16 commented:
Also, you may have a look about https://github.com/HomoArk/tracing-ohos. I just created this crate for redirecting `tracing` events/spans/logs to hilog. Their inner logic for them are somehow similar, which both leverage a Writer to print messages to hilog. 

Do we need to combine `tracing-ohos` and `hilog`?
